File: Advanced/BottomNavigationViewDemo/app/src/main/java/edu/cs4730/bottomnavigationviewdemo/MainActivity.java
package edu.cs4730.bottomnavigationviewdemo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import android.os.Bundle;
import android.view.MenuItem;

import com.google.android.material.badge.BadgeDrawable;
import com.google.android.material.bottomnavigation.BottomNavigationView;

/**
 * and example use the BottomNavigationView and Navigation.
 * https://developer.android.com/reference/com/google/android/material/bottomnavigation/BottomNavigationView
 * https://material.io/develop/android/components/bottom-navigation-view/
 */
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        BottomNavigationView navView = findViewById(R.id.nav_view);

        /* if not using arch navigation, then you need to implement this.
        navView.setOnItemSelectedListener(
            new BottomNavigationView.OnItemSelectedListener() {

                public boolean onNavigationItemSelected(MenuItem item) {
                    //setup the fragments here.
                    int id = item.getItemId();
                    if (id == R.id.action_first) {
                        getSupportFragmentManager().beginTransaction().replace(R.id.container, new OneFragment()).commit();
                        item.setChecked(true);
                        return true;
                    } else if (id == R.id.action_second) {
                        getSupportFragmentManager().beginTransaction().replace(R.id.container, new TwoFragment()).commit();
                        item.setChecked(true);
                    } else if (id == R.id.action_third) {
                        getSupportFragmentManager().beginTransaction().replace(R.id.container, new threeFragment()).commit();
                        item.setChecked(true);
                    }

                    return false;
                }
            }

        );*/


        // Passing each menu ID as a set of Ids because each menu should be considered as top level destinations.
        //Note for this to work with arch Navigation, these id must be the same id in menu.xml and the nav_graph.
        AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
            R.id.action_first, R.id.action_second, R.id.action_third)
            .build();
        //NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        NavHostFragment navHostFragment =
            (NavHostFragment) getSupportFragmentManager().findFragmentById(R.id.nav_host_fragment);
        NavController navController = navHostFragment.getNavController();

        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        NavigationUI.setupWithNavController(navView, navController);


        //In order to have badges, you need to use the Theme.MaterialComponents.DayNight  (doesn't have to be daynight, but MaterialComponents).
        BadgeDrawable badge = navView.getOrCreateBadge(R.id.action_second);
        badge.setNumber(12);  //should show a 12 in the "badge" for the second one.
        badge.setVisible(true);


    }
}

File: Basic/FragDemoSimple_kt/app/src/main/java/edu/cs4730/fragdemosimple_kt/MainActivity.kt
package edu.cs4730.fragdemosimple_kt

import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity

/**
 * This is a very simple demo of changing between two fragments.
 * This really not a good way to do this and is for demo purposes.  I needed something
 * really simple to start out with.
 */

class MainActivity : AppCompatActivity() {

    var firstfragment = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //if this a not new, then place add firstfragment to the framelayout
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction()
                .add(R.id.container, oneFragment())
                .commit()
        }
        //find the button and setup the listener.
        val btn1 = findViewById<Button>(R.id.button01)
        btn1.setOnClickListener {
            firstfragment = if (firstfragment) {
                //first fragment is showing, so replace it with the second one.
                supportFragmentManager.beginTransaction()
                    .replace(R.id.container, twoFragment())
                    .commit()
                false
            } else {
                //second fragment is showing, so replace it with the second one.
                supportFragmentManager.beginTransaction()
                    .replace(R.id.container, oneFragment())
                    .commit()
                true
            }
        }

    }
}